### PR TITLE
Add more `ref readonly` tests and address comments

### PR DESCRIPTION
**Pull request by jjonescz**
- Addressing pending comments from https://github.com/dotnet/roslyn/pull/68680
- Mentioning an issue per https://github.com/dotnet/roslyn/issues/69079#issuecomment-1643934930
- Adding tests for expression trees, operators, and IOperation

Test plan: https://github.com/dotnet/roslyn/issues/68056

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 1]
```diff
@@ -386,6 +386,9 @@ internal sealed class MemberSignatureComparer : IEqualityComparer<Symbol>
             _typeComparison = typeComparison;
             Debug.Assert((_typeComparison & TypeCompareKind.FunctionPointerRefMatchesOutInRefReadonly) == 0,
                          $"Rely on the {nameof(refKindCompareMode)} flag to set this to ensure all cases are handled.");
+            Debug.Assert((_refKindCompareMode & RefKindCompareMode.AllowRefReadonlyVsInMismatch) == 0 ||
+                (_refKindCompareMode & RefKindCompareMode.ConsiderDifferences) != 0,
+                $"Cannot set {nameof(RefKindCompareMode.AllowRefReadonlyVsInMismatch)} without {nameof(RefKindCompareMode.ConsiderDifferences)}.");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs:391]: Consider making the assert more general. I.e. either ```_refKindCompareMode == DoNotConsiderDifferences```, or the ```ConsiderDifferences``` flag is set.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 2]
```diff
@@ -628,6 +628,73 @@ static void verify(ModuleSymbol m)
...
+            // (5,37): error CS0631: ref and out are not valid in this context
+            //     public static explicit operator C(ref readonly int x) => null;
+            Diagnostic(ErrorCode.ERR_IllegalRefParam, "C").WithLocation(5, 37),
+            // (5,37): error CS0557: Duplicate user-defined conversion in type 'C'
+            //     public static explicit operator C(ref readonly int x) => null;
+            Diagnostic(ErrorCode.ERR_DuplicateConversionInClass, "C").WithArguments("C").WithLocation(5, 37));
+    }
+
+    [Fact]
+    public void ExpressionTrees()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:658]: >ExpressionTrees [](http://example.com/codeflow?start=16&length=15)

I think we also should test calling methods with `ref readonly` parameters inside an expression tree, With all possible call site modifiers, including none. Also, passing an rValue case.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 3]
```diff
@@ -4606,6 +4734,30 @@ static void Main()
             Diagnostic(ErrorCode.WRN_OverridingDifferentRefness, "M2").WithArguments("in int x", "ref readonly int x").WithLocation(9, 12));
     }
 
+    [Fact]
+    public void Implementation_RefReadonly_In_Close()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:4738]: >Implementation_RefReadonly_In_Close [](http://example.com/codeflow?start=16&length=35)

What is the goal of the test? Should we test success scenario instead by making the methods in ```C``` public?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5)

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs (hunk 3 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:4738]: No, the point is to test `MemberSignatureComparer.CSharpCloseImplicitImplementationComparer`

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8)

**Comment by jjonescz** [comment]
@jcouv for a second review, thanks

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 4]
```diff
@@ -783,6 +783,21 @@ private static void SubstituteConstraintTypes(ImmutableArray<TypeWithAnnotations
             }
 
             return true;
+
+            static bool areRefKindsCompatible(RefKindCompareMode refKindCompareMode, RefKind refKind1, RefKind refKind2)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs:787]: Thanks. This was much easier to follow :-)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 8)

**Pull request merged**